keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -111,153 +137,271 @@ Following conditions must be met:
-        * This is because kubelet / VolumeManager can track only one global volume mount (created by `NodeStageVolume`) and that mount can have only one SELinux context.
-          Volumes with different VolumeHandles have different global mounts and therefore can have different SELinux contexts.
-  * kubelet does not pass any special SELinux option to the container runtime (explicitly, no `:Z`).
-    Files on the volume already have the right SELinux context, no recursive relabeling happens there.
-  * When a `PodSecurityContext` specifies incomplete SELinux label (i.e. omits `SELinuxOptions.User`, `.Role` or `.Type`), kubelet fills the blanks from the system defaults provided by [ContainerLabels() from go-selinux bindings](https://github.com/opencontainers/selinux/blob/621ca21a5218df44259bf5d7a6ee70d720a661d5/go-selinux/selinux_linux.go#L770).
+* Allow CSI drivers to announce if they support mounting with `-o context` mount option via `CSIDriver spec.SELinuxMount bool` field.
+  * Under `SELinuxMountReadWriteOncePod` feature gate, beta + enabled by default in 1.28.
+  * Default value is `false`, i.e. no support for SELinux mount. It needs explicit opt-in.
+
+
+* Update kubelet to mount all volumes with `-o context` mount option, unless a Pod explicitly opts out.
+  * **Phase 1**: limited to ReadWriteOncePod (RWOP) volumes, under `SELinuxMountReadWriteOncePod` feature gate. RWOP volumes can't be used by multiple Pods and thus this cannot break any workload that mixes usage of the same volume by multiple Pods in parallel.
+  * **Phase 2**: same as Phase 1, the feature is still limited to RWOP volumes.
+    This phase enables opt-out that becomes active in Phase 3.
+  * **Phase 3**: all volumes are mounted, under `SELinuxMount` feature gate. It can break existing applications that need to mix privileged and unprivileged Pods using the same volume in parallel.
+    * We propose to send Pod events to immediately show why such Pods are stuck `ContainerCreating`.
+    * We propose metrics to identify such Pods before the `SELinuxMount` feature gate is enabled, to identify potential issues before a cluster upgrade / enabling the feature gate.
+    * We propose metrics to identify such Pods _after_ the `SELinuxMount` feature gate is enabled, to identify pods that are running only because they run on different nodes. If they landed on the same node, one of them would be stuck `ContainerCreating`, because they mix privileged and unprivileged pods.
+
+* The opt-out is realized by a new Pod field `PodSpec.SecurityContext.SELinuxChangePolicy` with values `MountOption` and `Recursive` (opt-out). `null` means `MountOption`.
+  * We need the field to be available in a cluster **before** `SELinuxMount` feature gate gets enabled, so cluster admins can fix their Pods and add opt-out before they upgrade to a version with `SELinuxMount` feature gate enabled.
```

Comment:
it'll be workload authors making the change, so both a metric and an event would help.

Reply:
Added a SELinuxController to kube-controller-manager to provide both events and metrics (so no kube-state-metrics changes are needed).